lib/TypeNames/ModelToPTMLTypeHelpers.cpp

```diff
@@ -589,136 +529,133 @@ void printDeclaration(Logger<> &Log,
                       ptml::PTMLCBuilder &B,
                       const model::Binary &Model,
-                      QualifiedTypeNameMap &AdditionalNames,
-                      const TypeSet &TypesToInline,
-                      llvm::StringRef NameOfInlineInstance,
-                      const std::vector<model::Qualifier> &Qualifiers,
-                      bool ForEditing) {
+                      TypeNameMap &AdditionalNames) {
   if (Log.isEnabled()) {
     auto Scope = helpers::LineComment(Header, B.isGenerateTagLessPTML());
     Header << "Declaration of " << getNameFromYAMLScalar(T.key()) << "\n";
   }
 
   revng_log(Log, "Declaring " << getNameFromYAMLScalar(T.key()));
 
-  switch (T.Kind()) {
+  if (auto *Enum = llvm::dyn_cast<model::EnumDefinition>(&T))
+    printForwardDeclaration(*Enum, Header, B);
 
-  case model::TypeDefinitionKind::Invalid: {
-    if (Log.isEnabled())
-      Header << B.getLineComment("invalid");
-  } break;
+  else if (auto *Struct = llvm::dyn_cast<model::StructDefinition>(&T))
+    printForwardDeclaration(*Struct, Header, B);
 
-  case model::TypeDefinitionKind::PrimitiveDefinition: {
-    // Do nothing. Primitive type declarations are all present in
-    // revng-primitive-types.h
-  } break;
+  else if (auto *Union = llvm::dyn_cast<model::UnionDefinition>(&T))
+    printForwardDeclaration(*Union, Header, B);
 
-  case model::TypeDefinitionKind::EnumDefinition: {
-    printForwardDeclaration(llvm::cast<model::EnumDefinition>(T), Header, B);
-  } break;
+  else if (auto *Typedef = llvm::dyn_cast<model::TypedefDefinition>(&T))
+    printDeclaration(*Typedef, Header, B);
 
-  case model::TypeDefinitionKind::StructDefinition: {
-    printForwardDeclaration(llvm::cast<model::StructDefinition>(T), Header, B);
-  } break;
+  else if (auto *RFD = llvm::dyn_cast<model::RawFunctionDefinition>(&T))
+    printDeclaration(Log, *RFD, Header, B, Model);
 
-  case model::TypeDefinitionKind::UnionDefinition: {
-    printForwardDeclaration(llvm::cast<model::UnionDefinition>(T), Header, B);
-  } break;
-
-  case model::TypeDefinitionKind::TypedefDefinition: {
-    printDeclaration(llvm::cast<model::TypedefDefinition>(T), Header, B);
-  } break;
-
-  case model::TypeDefinitionKind::RawFunctionDefinition: {
-    printDeclaration(Log,
-                     llvm::cast<model::RawFunctionDefinition>(T),
-                     Header,
-                     B,
-                     Model);
-  } break;
-
-  case model::TypeDefinitionKind::CABIFunctionDefinition: {
-    printDeclaration(llvm::cast<model::CABIFunctionDefinition>(T),
-                     Header,
-                     B,
-                     AdditionalNames,
-                     Model);
-  } break;
-  default:
-    revng_abort();
-  }
+  else if (auto *CFD = llvm::dyn_cast<model::CABIFunctionDefinition>(&T))
+    printDeclaration(*CFD, Header, B, AdditionalNames, Model);
+
+  else
+    revng_abort("Unsupported type definition.");
```

Comment:
Can't we have a `switch` here? All those if-else-if hurt my eyes.

Reply:
The thing I don't like about switches is that they require a separate cast inside, so two-liner like
```cpp
  else if (auto *CFD = llvm::dyn_cast<model::CABIFunctionDefinition>(&T))
    printDeclaration(*CFD, Header, B, AdditionalNames, Model);
```
suddenly twice the length:
```cpp
  case model::TypeDefinitionKind::CABIFunctionDefinition: {
    auto &CFD = llvm::cast<model::CABIFunctionDefinition>(&T);
    printDeclaration(CFD, Header, B, AdditionalNames, Model);
  }
```

Reply:
I was gonna say we have `upcast` for that.

But then the various `printDeclaration` should have a matching prototype, which unfortunately isn't the case.

Forget about this.

Reply:
Yeah, and also as far as I remember, `upcast` is not friends with coroutines either, which can make it even more complicated.